apps/champions/lib/champions/battle/simulator.ex

```diff
@@ -63,187 +64,459 @@ defmodule Champions.Battle.Simulator do
+  # If the effect hasn't triggered yet, reduce the remaining effects_trigger counter.
+  defp process_skill_effects_trigger_value(skill, current_state, _initial_step_state) do
+    {%{skill | effects_trigger: skill.effects_trigger - 1}, current_state}
+  end
 
-    Enum.reduce(skill.effects, current_state, fn effect, new_state ->
-      # If skill doesn't have targeting strategy, we fall back to the effects'
-      target_ids =
-        if is_nil(target_ids),
-          do: choose_targets(unit, effect, initial_step_state),
-          else: target_ids
+  # Calculate the new state of the battle after a step passes for a pending effect.
+  # If the effect is ready to be processed, we apply it.
+  defp process_step_for_effect(%{delay: 0} = effect, current_state) do
+    Logger.info("#{format_unit_name(effect.caster)}'s effect is ready to be processed")
 
-      targets_after_effect = Enum.map(target_ids, fn id -> apply_effect(effect, unit, new_state[id]) end)
+    targets_after_effect =
+      Enum.map(effect.targets, fn id ->
+        maybe_apply_effect(effect, current_state.units[id], effect.caster, current_state.step_number)
+      end)
 
-      Enum.reduce(targets_after_effect, new_state, fn target_unit, acc_state ->
-        Map.put(acc_state, target_unit.id, target_unit)
+    current_state =
+      Enum.reduce(targets_after_effect, current_state, fn target, acc_state ->
+        put_in(acc_state, [:units, target.id], target)
       end)
-    end)
+
+    Map.put(current_state, :pending_effects, List.delete(current_state.pending_effects, effect))
+  end
+
+  # Calculate the new state of the battle after a step passes for a pending effect.
+  # If the effect isn't ready to be processed, we reduce its remaining delay.
+  defp process_step_for_effect(effect, current_state) do
+    Map.put(current_state, :pending_effects, [
+      %{effect | delay: effect.delay - 1} | List.delete(current_state.pending_effects, effect)
+    ])
   end
 
-  defp choose_targets(_unit, %{targeting_strategy: nil}, _state), do: nil
+  # Check if the unit can attack this turn.
+  # For now, attacking capability is only affected by whether the unit is currently casting a skill.
+  # Later on, things like stuns will be handled here.
+  defp can_attack(unit, initial_step_state) do
+    # Check the unit is not casting anything right now
+    not Enum.any?(initial_step_state.skills_being_cast, &(&1.caster_id == unit.id))
+  end
+
+  # Check if the unit can cast their ultimate skill this step.
+  defp can_cast_ultimate_skill(unit), do: unit.energy >= @ultimate_energy_cost
+
+  # Check if the unit can cast their basic skill this step.
+  defp can_cast_basic_skill(unit), do: unit.basic_skill.remaining_cooldown <= 0
+
+  # Called when a skill being cast reaches effect_trigger 0.
+  # "Queues" the effect to be processed when its delay reaches 0.
+  defp trigger_skill_effects(skill, current_state, initial_step_state) do
```

Comment:
i dont think the name "trigger_skill_effect" is suitable for what this method is doing, it leads me to think that we'll change the unit states with the skill effects but we're just updating the effect and adding it to the pending effects list, maybe `queue_skill_effects` or something like that would be more suitable

Reply:
I've intentionally avoided calling it "queue" since it's not technically a queue, as processing depends on the delay and not on order of creation. I chose this name because the configuration name the game designe asked for was "animation_trigger", so I left trigger as the verb, but I can see why one could confuse it with application. Can you think of a third alternative?

Reply:
Hmm we could do something like `add_skill_effects_to_pending_effects` or `append_skill_effects_for_application`

Reply:
add_skill_effects_to_pending_effects works just fine!